File: backend/src/main/java/com/smartcampus/service/BookingService.java
package com.smartcampus.service;

import com.smartcampus.dto.BookingListItemDTO;
import com.smartcampus.entity.User;
import com.smartcampus.model.Booking;
import com.smartcampus.model.Notification;
import com.smartcampus.repository.BookingRepository;
import com.smartcampus.repository.UserRepository;
import com.smartcampus.security.Role;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;

/**
 * Booking Service
 * Handles business logic for booking management
 */
@Service
@Transactional
@Slf4j
public class BookingService {

    @Autowired
    private BookingRepository bookingRepository;

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private NotificationService notificationService;

    public Booking createBooking(Booking booking) {
        Booking savedBooking = bookingRepository.save(booking);
        
        // Create notification for all admins
        notifyAdminsAboutBooking(savedBooking);
        
        return savedBooking;
    }

    private void notifyAdminsAboutBooking(Booking booking) {
        try {
            List<User> admins = userRepository.findByRole(Role.ADMIN);
            for (User admin : admins) {
                String title = "New Booking Request from " + (booking.getUser() != null ? booking.getUser().getName() : "Unknown");
                String resourceName = booking.getResource() != null ? booking.getResource().getName() : "Unknown Resource";
                String bookingDate = booking.getBookingDate() != null ? booking.getBookingDate().toString() : "TBD";
                String description = "Booking for " + resourceName + " on " + bookingDate;
                
                notificationService.createNotification(
                    admin,
                    title,
                    description,
                    Notification.NotificationType.BOOKING,
                    Notification.NotificationSeverity.INFO
                );
            }
            log.info("Created booking notifications for {} admins", admins.size());
        } catch (Exception e) {
            log.error("Error creating booking notification", e);
        }
    }

    public Booking getBookingById(Long id) {
        return bookingRepository.findById(id).orElse(null);
    }

    public List<Booking> getAllBookings() {
        return bookingRepository.findAll();
    }

    public List<Booking> getBookingsByResource(Long resourceId) {
        return bookingRepository.findByResourceId(resourceId);
    }

    public List<Booking> getBookingsByUser(Long userId) {
        return bookingRepository.findByUserId(userId);
    }

    /**
     * Lists bookings via JDBC so Hibernate 6 / MySQL 5.5 quirks (join fetch, time mapping) cannot 500 the UI.
     */
    @Transactional(readOnly = true)
    public List<BookingListItemDTO> listBookingsForUser(Long userId) {
        final String sql = """
                SELECT b.id,
                       b.resource_id,
                       r.name AS res_name,
                       r.location AS res_location,
                       b.user_id,
                       u.name AS user_name,
                       u.email AS user_email,
                       b.booking_date,
                       b.start_time,
                       b.end_time,
                       b.purpose,
                       b.expected_attendees,
                       b.status,
                       b.admin_comment,
                       b.created_at
                FROM bookings b
                LEFT JOIN resources r ON r.id = b.resource_id
                LEFT JOIN users u ON u.id = b.user_id
                WHERE b.user_id = ?
                ORDER BY b.booking_date DESC, b.start_time DESC
                """;
        return jdbcTemplate.query(sql, (rs, rowNum) -> BookingListItemDTO.fromJoinedRow(rs), userId);
    }

    @Transactional(readOnly = true)
    public List<BookingListItemDTO> listAllBookings() {
        final String sql = """
                SELECT b.id,
                       b.resource_id,
                       r.name AS res_name,
                       r.location AS res_location,
                       b.user_id,
                       u.name AS user_name,
                       u.email AS user_email,
                       b.booking_date,
                       b.start_time,
                       b.end_time,
                       b.purpose,
                       b.expected_attendees,
                       b.status,
                       b.admin_comment,
                       b.created_at
                FROM bookings b
                LEFT JOIN resources r ON r.id = b.resource_id
                LEFT JOIN users u ON u.id = b.user_id
                ORDER BY b.booking_date DESC, b.start_time DESC
                """;
        return jdbcTemplate.query(sql, (rs, rowNum) -> BookingListItemDTO.fromJoinedRow(rs));
    }

    @Transactional(readOnly = true)
    public BookingListItemDTO getBookingListItem(Long bookingId) {
        final String sql = """
                SELECT b.id,
                       b.resource_id,
                       r.name AS res_name,
                       r.location AS res_location,
                       b.user_id,
                       u.name AS user_name,
                       u.email AS user_email,
                       b.booking_date,
                       b.start_time,
                       b.end_time,
                       b.purpose,
                       b.expected_attendees,
                       b.status,
                       b.admin_comment,
                       b.created_at
                FROM bookings b
                LEFT JOIN resources r ON r.id = b.resource_id
                LEFT JOIN users u ON u.id = b.user_id
                WHERE b.id = ?
                """;
        List<BookingListItemDTO> rows =
                jdbcTemplate.query(sql, (rs, rowNum) -> BookingListItemDTO.fromJoinedRow(rs), bookingId);
        return rows.isEmpty() ? null : rows.get(0);
    }

    public List<Booking> getBookingsByStatus(String status) {
        return bookingRepository.findByStatus(status);
    }

    public List<Booking> getBookingsByDateRange(LocalDate startDate, LocalDate endDate) {
        return bookingRepository.findByBookingDateBetween(startDate, endDate);
    }

    public Booking updateBooking(Booking booking) {
        return bookingRepository.save(booking);
    }

    public void deleteBooking(Long id) {
        bookingRepository.deleteById(id);
    }

    public boolean approveBooking(Long id, String adminComment) {
        Booking booking = bookingRepository.findById(id).orElse(null);
        if (booking != null) {
            booking.setStatus("APPROVED");
            booking.setAdminComment(adminComment);
            bookingRepository.save(booking);
            return true;
        }
        return false;
    }

    public boolean rejectBooking(Long id, String adminComment) {
        Booking booking = bookingRepository.findById(id).orElse(null);
        if (booking != null) {
            booking.setStatus("REJECTED");
            booking.setAdminComment(adminComment);
            bookingRepository.save(booking);
            return true;
        }
        return false;
    }
}

File: frontend/src/services/bookingService.ts
import { apiClient } from './axiosConfig'

const API_URL = '/api/bookings'

export interface BookingCreateRequest {
  resourceId: number
  bookingDate: string
  startTime: string
  endTime: string
  purpose: string
  expectedAttendees: number
}

export interface BookingResponse {
  id: number
  resourceId: number
  resourceName: string
  resourceLocation: string
  userId?: number
  userName?: string
  userEmail?: string
  bookingDate: string
  startTime: string
  endTime: string
  purpose: string
  expectedAttendees: number
  status: string
  adminComment?: string | null
  createdAt: string
}

export interface BookingAdminFilters {
  status?: string
  startDate?: string
  endDate?: string
  resourceId?: number
  userId?: number
}

export const createBooking = (payload: BookingCreateRequest) =>
  apiClient.post<BookingResponse>(API_URL, payload)

export const getMyBookings = () =>
  apiClient.get<BookingResponse[]>(`${API_URL}/my`)

export const getAllBookings = (filters?: BookingAdminFilters) => {
  const params = new URLSearchParams()
  if (filters?.status) params.append('status', filters.status)
  if (filters?.startDate) params.append('startDate', filters.startDate)
  if (filters?.endDate) params.append('endDate', filters.endDate)
  if (typeof filters?.resourceId === 'number' && Number.isFinite(filters.resourceId)) {
    params.append('resourceId', String(filters.resourceId))
  }
  if (typeof filters?.userId === 'number' && Number.isFinite(filters.userId)) {
    params.append('userId', String(filters.userId))
  }
  const query = params.toString()
  return apiClient.get<BookingResponse[]>(query ? `${API_URL}?${query}` : API_URL)
}

export const updateBooking = (id: number, payload: BookingCreateRequest) =>
  apiClient.put<BookingResponse>(`${API_URL}/${id}`, payload)

export const deleteBooking = (id: number) =>
  apiClient.delete(`${API_URL}/${id}`)

export const cancelBooking = (id: number) =>
  apiClient.patch<BookingResponse>(`${API_URL}/${id}/cancel`)

export const approveBooking = (id: number) =>
  apiClient.patch<BookingResponse>(`${API_URL}/${id}/approve`)

export const rejectBooking = (id: number, reason: string) =>
  apiClient.patch<BookingResponse>(`${API_URL}/${id}/reject`, { reason })
